src/oauth.tsx

```diff
@@ -218,35 +275,34 @@ app.post("/token", cors(), clientAuthentication, async (c) => {
             accessGrant.applicationId !== client.id ||
             accessGrant?.revoked !== null
           ) {
-            return c.json(
-              {
-                error: "invalid_grant",
-                error_description: INVALID_GRANT_ERROR_DESCRIPTION,
-              },
-              400,
+            return c.json(INVALID_GRANT_ERROR, 400);
+          }
+
+          if (accessGrant.codeChallenge && accessGrant.codeChallengeMethod) {
+            if (
+              !form.code_verifier ||
+              accessGrant.codeChallengeMethod !== "S256"
```

Comment:
@dahlia there's now an explicit check here that the stored `codeChallengeMethod` is one that we actually support.